juju/juju
# Properly reuse existing devices 1782338

## Description of change

When fixing [bug #1670873](pad.lv/1670873), we made it possible for the 'retry' logic of AllocateContainerAddresses to re-use a device that might have been created in the first attempt. However, we only really tested that support with containers that had a single interface. And it turned out our logic around reusing the existing device would do the wrong thing if we wanted multiple interfaces (it would assume it still needed to create all but the primary interface).

This refactors the code quite a bit, but mostly in an attempt to make it easier to understand the logical steps of how we create a device, and populate all of its interfaces.
The main functional change is that we'll now look for an existing device, and then validate if the existing device matches all of the requested network information. If it does, then we will use it, if it doesn't, then we will Delete() it and create it as though it didn't exist.

## QA steps

Ideally we would have a way to inject failure after the point of device registration in MAAS. I don't have amazing ideas here, something around causing LXD to fail to provision, injecting some sort of code that will fail once, but not every time, etc. We should see that containers with multiple network interfaces still continue to be provisioned. We might also be able to test it by allocating a MAAS machine, and then manually issuing a "CreateDevice" call with the name we know we are going to give a container (the name we ask MAAS to use matches the machine id in Juju). We should see that it doesn't use the device that is improperly configured and Deletes it to add the right definition.

## Documentation changes

No

## Bug reference

[lp:1682338](https://bugs.launchpad.net/juju/+bug/1682338)
